examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,279 @@
+ */
+const NUM_CPUS: usize = 2;
+/* PER_THREAD_PER_BUFFER_MAX_SIZE is max size of a buffer that a thread can allocate
+ * on each call to add_new_buffer() function. Higher the value, more will be
+ * the total memory consumption and more time taken by the test.
+ */
+const PER_THREAD_PER_BUFFER_MAX_SIZE: usize = 4 * TO_KB;
+/* MAX_BUFFER_CHECKS_PER_THREAD_ITERATION is the maximum number of buffers to
+ * check on each call to select_and_check_random_buffer_contents()
+ */
+const MAX_BUFFER_CHECKS_PER_THREAD_ITERATION: usize = 8;
+/* MAX_INDEX_CHECKS_PER_BUFFER is the number of indices/locations to check
+ * per buffer.
+ */
+const MAX_INDEX_CHECKS_PER_BUFFER: usize = 1 << 16;
+/* MIN_ALLOWED_FREE_HEAP_PERCENTAGE is the fraction free memory below which
+ * we will allow the vector allocations to fail
+ */
+const MIN_ALLOWED_FREE_HEAP_PERCENTAGE: f64 = 10.0;
+/* MAX_ITERATIONS_PER_THREAD is the number of operations per thread (1 operation per
+ * per iteration)
+ */
+const MAX_ITERATIONS_PER_THREAD: usize = 32;
+
+#[inline(always)]
+pub fn image_base() -> u64 {
+    let base: u64;
+    unsafe {
+        asm!(
+            "lea IMAGE_BASE(%rip), {}",
+            lateout(reg) base,
+            options(att_syntax, nostack, preserves_flags, nomem, pure),
+        )
+    };
+    base
+}
+
+#[inline(always)]
+pub(crate) unsafe fn rel_ptr_mut<T>(offset: u64) -> *mut T {
+    (image_base() + offset) as *mut T
+}
+
+/* Returns the base memory address of the heap */
+pub(crate) fn heap_base() -> *const u8 {
+    unsafe { rel_ptr_mut(HEAP_BASE) }
+}
+
+/* Returns the size of the heap */
+pub(crate) fn heap_size() -> usize {
+    unsafe { HEAP_SIZE }
+}
+
+fn update_occupied_heap_size_on_delete(size: usize) {
+    HEAP_ALLOCATED.fetch_sub(size, Ordering::SeqCst);
+}
+
+fn update_occupied_heap_size_on_addition(size: usize) {
+    HEAP_ALLOCATED.fetch_add(size, Ordering::SeqCst);
+}
+
+fn get_occupied_heap_size() -> usize {
+    HEAP_ALLOCATED.load(Ordering::SeqCst)
+}
+
+fn get_free_heap_size_in_bytes() -> usize {
+    heap_size() - get_occupied_heap_size()
+}
+
+fn get_free_heap_percentage() -> f64 {
+    (get_free_heap_size_in_bytes() as f64 / heap_size() as f64) * 100.0
+}
+
+fn get_random_num(start: usize, end: usize) -> usize {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(start..=end)
+}
+
+fn wait_per_thread(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    while !*started {
+        started = cvar.wait(started).unwrap();
+    }
+    drop(started);
+}
+
+fn wakeup_all_child_threads(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    *started = true;
+    cvar.notify_all();
+    drop(started);
+}
+
+fn traverse_and_check_buffer(buf: &Vec<usize>, magic_num_to_check: usize) {
+    let num_indices_checks = get_random_num(1, MAX_INDEX_CHECKS_PER_BUFFER);
```

Comment:
I think you have the same issue here as in #566: You allocate a `Vec` assuming each element is a byte, which is not the case